pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+	// If daysToLive is zero, use the default.
+	if daysToLive == 0 {
+		daysToLive = defaultDaysToLive
+	}
+	s := strconv.FormatUint(24*daysToLive, 10) + "h"
+	expirationAge, err := time.ParseDuration(s)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid expirationDays value configured: %v", err)
+	}
+	s = strconv.FormatUint(12*daysToLive, 10) + "h" // half of the expiration age
+	refUpdateThresh, _ := time.ParseDuration(s)
+
+	spannerBlobAccessPrometheusMetrics.Do(func() {
+		prometheus.MustRegister(spannerMalformedKeyCount)
+		prometheus.MustRegister(gcsReftimeUpdateCount)
+		prometheus.MustRegister(gcsReftimeUpdateFailedCount)
+		prometheus.MustRegister(spannerReftimeUpdateCount)
+		prometheus.MustRegister(spannerReftimeUpdateFailedCount)
+		prometheus.MustRegister(spannerMalformedBlobDeletedCount)
+		prometheus.MustRegister(spannerMalformedBlobDeleteFailedCount)
+		prometheus.MustRegister(spannerExpiredBlobReadIgnoredCount)
+		prometheus.MustRegister(gcsFailedReadDeletedBlobCount)
+		prometheus.MustRegister(gcsFailedReadDeleteBlobFailedCount)
+		prometheus.MustRegister(gcsPutFailedContextCanceledCount)
+		prometheus.MustRegister(gcsPutFailedDeadlineExceededCount)
+		prometheus.MustRegister(gcsPutFailedOtherCount)
+		prometheus.MustRegister(backendOperationsDurationSeconds)
+	})
+
+	ctx := context.Background()
+	spannerClient, err := spanner.NewClient(ctx, databaseName)
+	if err != nil {
+		log.Printf("Can't create spanner client: %v", err)
+		return nil, err
+	}
+
+	// If the spanner table doesn't exist, create it.
+	stmt := spanner.NewStatement(`SELECT 1 FROM information_schema.tables WHERE table_name = "` + tableName + `"`)
+	iter := spannerClient.Single().Query(ctx, stmt)
+	_, err = iter.Next()
+	iter.Stop()
+	log.Printf("spanner table check, err = %v", err)
+	if err == iterator.Done {
+		// Table doesn't exist.  Create it.
+		if err = createSpannerTable(ctx, databaseName, daysToLive); err != nil {
+			// We could have raced with another pod.  Check if the table exists.
+			iter := spannerClient.Single().Query(ctx, stmt)
+			_, xerr := iter.Next()
+			iter.Stop()
+			if xerr == iterator.Done {
+				// Table still doesn't exist.
+				spannerClient.Close()
+				log.Printf("Can't create spanner table: %v", err)
+				return nil, err
+			}
+		}
+	} else if err == nil {
+		// Check if we need to update the TTL.
+		days, err := getSpannerTTL(ctx, databaseName)
+		if err != nil {
+			log.Printf("Can't determine Spanner TTL: %v", err)
+		}
+		if days != daysToLive {
+			err = updateSpannerDeletionPolicy(ctx, databaseName, daysToLive)
+			if err != nil {
+				log.Printf("Can't update Spanner TTL: %v", err)
+			} else {
+				log.Printf("Spanner TTL changed from %d to %d days", days, daysToLive)
+			}
+		}
+	}
+
+	storageClient, err := storage.NewClient(ctx)
+	if err != nil {
+		spannerClient.Close()
+		log.Printf("Can't create GCS client: %v", err)
```

Comment:
Either log errors or propagate them, but not both.

Reply:
That only works if the caller receiving the error logs/prints it.  In this case, it's enabling us to debug start-up problems by simply looking at the k8s logs.

Reply:
You can also install things like middleware for gRPC to do the logging. That way we don't need to plaster our codebase with logging statements.